tests/31sync/06state.pl

```diff
@@ -745,3 +745,158 @@ sub wait_for_event_in_room {
          Future->done(1);
       })
    };
+
+# Test to check that current state events appear in the timeline,
+# even if they were set during a period the user couldn't see
+test "Current state appears in timeline in private history",
+   requires => [ local_user_fixtures( 3, with_events => 0 ),
+                 qw( can_sync ) ],
+
+   check => sub {
+      my ( $creator, $syncer, $invitee ) = @_;
+
+      my ( $room_id );
+
+      matrix_create_room( $creator )
+      ->then( sub {
+         ( $room_id ) = @_;
+
+         matrix_join_room( $syncer, $room_id )
+      })->then( sub {
+         matrix_set_room_history_visibility( $creator, $room_id, "joined" )
+      })->then( sub {
+         matrix_invite_user_to_room( $creator, $invitee, $room_id )
+      })->then( sub {
+         matrix_sync( $syncer )
+      })->then( sub {
+         matrix_leave_room( $syncer, $room_id )
+      })->then( sub {
+         matrix_join_room( $invitee, $room_id )
+      })->then( sub {
+         matrix_join_room_synced( $syncer, $room_id )
+      })->then( sub {
+         matrix_sync_again( $syncer )
+      })->then( sub {
+         my ( $body ) = @_;
+
+         my $room = $body->{rooms}{join}{$room_id};
+
+         # Check that we see invitee join event
+         any {
+            $_->{type} eq "m.room.member"
+            && $_->{state_key} eq $invitee->user_id
+            && $_->{content}{membership} eq "join"
+         } @{ $room->{timeline}{events} }
```

Comment:
not sure I'd expect this to appear in the timeline, but ymmv

Reply:
It can't really be put in the "state" part, as that specifically is for state at the start of the timeline, which would rightly have the "invite" in. It does feel quite icky though :/